File: game-server/src/main/java/net/dodian/uber/game/networking/game/GamePacketDecoder.java
package net.dodian.uber.game.networking.game;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ByteToMessageDecoder;
import io.netty.util.AttributeKey;
import net.dodian.utilities.ISAACCipher;
import net.dodian.uber.game.Constants;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

/**
 * Zero-copy Netty decoder that converts the raw RS317 stream into {@link GamePacket}s.
 * Logic mirrors {@link net.dodian.uber.game.networking.game.IsaacDecoder} but keeps the
 * payload inside a {@link ByteBuf} slice instead of copying into a byte[].
 */
public class GamePacketDecoder extends ByteToMessageDecoder {

    private static final Logger logger = LoggerFactory.getLogger(GamePacketDecoder.class);

    private static final int VARIABLE_BYTE  = -1;
    private static final int VARIABLE_SHORT = -2;

    private static final AttributeKey<ISAACCipher> IN_CIPHER_KEY = AttributeKey.valueOf("inCipher");

    private int opcode = -1;
    private int size   = 0;

    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
        ISAACCipher cipher = ctx.channel().attr(IN_CIPHER_KEY).get();
        if (cipher == null) {
            logger.warn("[Netty] Missing ISAAC cipher attribute – closing {}", ctx.channel().remoteAddress());
            ctx.close();
            return;
        }

        while (true) {
            if (opcode == -1) {
                if (!in.isReadable()) {
                    return; // need more data
                }
                opcode = (in.readUnsignedByte() - cipher.getNextKey()) & 0xFF;
                if (opcode < 0 || opcode >= Constants.PACKET_SIZES.length) {
                    logger.debug("[Netty] Invalid packet opcode {} from {}", opcode, ctx.channel().remoteAddress());
                    ctx.close();
                    return;
                }
                size = Constants.PACKET_SIZES[opcode];
            }

            if (size == VARIABLE_BYTE) {
                if (in.readableBytes() < 1) {
                    return;
                }
                size = in.readUnsignedByte();
            } else if (size == VARIABLE_SHORT) {
                if (in.readableBytes() < 2) {
                    return;
                }
                size = in.readUnsignedShort();
            }

            if (in.readableBytes() < size) {
                return; // wait for full payload
            }

            // Slice without copy
            ByteBuf payload = in.readSlice(size).retain();
            out.add(new GamePacket(opcode, size, payload));

            // reset state for next packet
            opcode = -1;
            size   = 0;

            if (!in.isReadable()) {
                return;
            }
        }
    }
}

File: game-server/src/main/kotlin/net/dodian/uber/net/codec/game/GamePacketDecoder.kt
package net.dodian.uber.net.codec.game

import com.github.michaelbull.logging.InlineLogger
import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import io.netty.channel.ChannelHandlerContext
import net.dodian.uber.net.message.GameProtocol
import net.dodian.uber.net.message.meta.PacketType
import net.dodian.uber.utils.StatefulFrameDecoder
import net.dodian.utilities.security.IsaacRandom

private val logger = InlineLogger()

class GamePacketDecoder(
    private val random: IsaacRandom,
    private val protocol: GameProtocol
) : StatefulFrameDecoder<GameDecoderState>(GameDecoderState.GAME_OPCODE) {
    var length: Int = 0
    var opcode: Int = -1
    var type: PacketType = PacketType.FIXED

    override fun decode(ctx: ChannelHandlerContext, input: ByteBuf, out: MutableList<Any>, state: GameDecoderState) {
        when (state) {
            GameDecoderState.GAME_OPCODE -> decodeOpcode(input, out)
            GameDecoderState.GAME_LENGTH -> decodeLength(input)
            GameDecoderState.GAME_PAYLOAD -> decodePayload(input, out)
            else -> error("Illegal decoder state: $state")
        }
    }

    private fun decodeOpcode(buffer: ByteBuf, out: MutableList<Any>) {
        if (!buffer.isReadable) return

        val encryptedOpcode = buffer.readUnsignedByte().toInt()
        opcode = encryptedOpcode - random.nextInt() and 0xFF

        val metadata = protocol.metadata[opcode]
            ?: error("No metadata for opcode: $opcode")

        type = metadata.type
        when (type) {
            PacketType.FIXED -> {
                length = metadata.length
                if (length == 0) {
                    state = GameDecoderState.GAME_OPCODE
                    out.add(GamePacket(opcode, type, Unpooled.EMPTY_BUFFER))
                } else state = GameDecoderState.GAME_PAYLOAD
            }
            PacketType.VARIABLE_BYTE -> state = GameDecoderState.GAME_LENGTH
            else -> error("Illegal packet type: $type")
        }
    }

    private fun decodeLength(buffer: ByteBuf) {
        if (!buffer.isReadable) return

        length = buffer.readUnsignedByte().toInt()
        if (length != 0) state = GameDecoderState.GAME_PAYLOAD
    }

    private fun decodePayload(buffer: ByteBuf, out: MutableList<Any>) {
        if (buffer.readableBytes() < length) return

        val payload = buffer.readBytes(length)
        state = GameDecoderState.GAME_OPCODE
        out.add(GamePacket(opcode, type, payload))
    }
}